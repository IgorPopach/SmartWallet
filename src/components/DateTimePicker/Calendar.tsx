import React from 'react';
import { parse } from '../../services/time';

interface Props {
    data: number;
    updateData: (value: number) => void;
}

const MONTHS = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

const DAYS_SHORT = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

const Calendar = ({ data, updateData }: Props) => {
    const parsedData = React.useMemo(() => parse(data), []);

    const [selectedDate, setSelectedDate] = React.useState(parsedData);

    const [date, setDate] = React.useState(selectedDate);

    const month = React.useMemo(() => date.month, [date]);
    const year = React.useMemo(() => date.year, [date]);

    const days = React.useMemo(() => {
        const offset = date.set({ day: 1 }).weekday;
        const daysArr = new Array(0).concat(new Array(offset - 1).fill(null));
        new Array(date.daysInMonth).fill(null).map((_, index) => daysArr.push(++index));
        return daysArr;
    }, [date]);

    const weeks = React.useMemo(
        () =>
            new Array(Math.ceil(days.length / 7)).fill(null).map((_, index) => days.slice(index * 7, (index + 1) * 7)),
        [date],
    );

    const tableBody = React.useMemo(
        () =>
            weeks.map((week, index) => (
                <tr key={index}>
                    {week.map((d, dIndex) => {
                        let styles = 'calendar-day';
                        if (d === null) {
                            return <td key={`${index} - ${dIndex}`}>{d}</td>;
                        }
                        if (d === selectedDate.day && month === selectedDate.month && year === selectedDate.year) {
                            styles = 'calendar-day is-selected';
                        }
                        const handleDayClick = () => {
                            setSelectedDate((prevState) => {
                                const updateDate = prevState.set({ day: d, month });
                                updateData(updateDate.toMillis());
                                return updateDate;
                            });
                        };
                        return (
                            <td key={`${index} - ${dIndex}`} className={styles} onClick={handleDayClick}>
                                {d}
                            </td>
                        );
                    })}
                </tr>
            )),
        [date],
    );

    const clickPrevMonth = React.useCallback(() => setDate((prevState) => prevState.minus({ month: 1 })), [date]);

    const clickNextMonth = React.useCallback(() => setDate((prevState) => prevState.plus({ month: 1 })), [date]);

    React.useEffect(() => {
        setDate(selectedDate);
    }, [selectedDate]);

    return (
        <>
            <div className="calendar-header">
                <div className="calendar-btn prev-month" onClick={clickPrevMonth}>
                    ◀
                </div>
                <div className="calendar-header-view">
                    {MONTHS[month - 1]} {year}
                </div>
                <div className="calendar-btn next-month" onClick={clickNextMonth}>
                    ▶
                </div>
            </div>
            <table className="calendar-body">
                <thead className="calendar-body__weeks">
                    <tr>
                        {DAYS_SHORT.map((d, id) => (
                            <th key={id}>{d}</th>
                        ))}
                    </tr>
                </thead>
                <tbody className="calendar-days">{tableBody}</tbody>
            </table>
        </>
    );
};

export default Calendar;
