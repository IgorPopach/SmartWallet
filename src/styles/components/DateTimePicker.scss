.datetime__frame {
  width: 310px;
  background-color: $secondary;
  color: $primary;
  border: 1px solid lighten($primary, 10%);
  border-radius: 3px;
  box-shadow: 2px 2px 2px rgb(148, 148, 148);
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 5px;
}

.datetime__control {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;

  .datetime_btn {
    border: 1px solid lighten($primary, 20%);
    width: 100%;
    text-align: center;
    cursor: pointer;

    &:first-child {
      border-top-left-radius: 3px;
      border-right: none;
    }

    &:last-child {
      border-top-right-radius: 3px;
    }
  }

  .is-active {
    background: lighten($primary, 30%);
    color: darken($secondary, 20%);
  }

}

.datetime__body {
  height: 200px;
  width: 100%;
  border: 1px solid $primary;
  border-top: none;
  border-bottom-left-radius: 3px;
  border-bottom-right-radius: 3px;
}

.datetime_submit_btn {
  margin: 5px 0 0;
  padding: 3px 15px;
  background-color: $primary;
  color: $secondary;
  border: 1px solid darken($primary, 20%);
  border-radius: 3px;
  transition: background-color .3s ease;
  
  &:hover {
    cursor: pointer;
    background-color: $dark;
  }
}