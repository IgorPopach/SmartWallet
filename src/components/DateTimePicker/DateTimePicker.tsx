import React from 'react';

import Calendar from './Calendar';
import Time from './Time';
import { now } from '../../services/time';
import { toClasses } from '../../utils';

interface Props {
    date?: number;
    handleClose?: () => void;
    onChange: (value: number) => void;
    className?: string;
}

const DateTimePicker = ({ date, handleClose, onChange, className }: Props) => {
    const [tab, setTab] = React.useState(0);
    const [data, setData] = React.useState(date || (now('number') as number));

    const styles = React.useMemo(() => toClasses(['datetime__frame'].concat(className)), [className]);

    const handleDateClick = React.useCallback(() => setTab(0), []);
    const handleTimeClick = React.useCallback(() => setTab(1), []);

    const dateButton = React.useMemo(() => {
        if (tab === 0) {
            return 'datetime_btn is-active';
        }
        return 'datetime_btn';
    }, [tab]);

    const timeButton = React.useMemo(() => {
        if (tab === 1) {
            return 'datetime_btn is-active';
        }
        return 'datetime_btn';
    }, [tab]);

    const updateData = React.useCallback((value) => {
        setData(value);
        return onChange(value);
    }, []);

    const toggleComponent = React.useMemo(() => {
        if (tab === 1) {
            return <Time {...{ data, updateData }} />;
        }
        return <Calendar {...{ data, updateData }} />;
    }, [tab, data]);

    // const handleSubmit = React.useCallback(() => {
    //   onChange(data);
    //   handleClose();
    // }, [data]);

    return (
        <div className={styles}>
            <div className="datetime__control">
                <div className={dateButton} {...{ updateData }} onClick={handleDateClick}>
                    Date
                </div>
                <div className={timeButton} onClick={handleTimeClick}>
                    Time
                </div>
            </div>
            <div className="datetime__body">{toggleComponent}</div>
            <div className="datetime_submit_btn" onClick={handleClose}>
                Set
            </div>
        </div>
    );
};

export default DateTimePicker;
