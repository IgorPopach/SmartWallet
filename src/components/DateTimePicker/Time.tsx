import React from 'react';
import { parse } from '../../services/time';
import TimeList from './TimeList';

export const TIME_TYPE = {
    HOUR: 'hour',
    MINUTE: 'minute',
};

interface Props {
    data: number;
    updateData: (value: number) => void;
}

export const timeIntoString = (value: number) => ('0' + value).slice(-2);

const Time = ({ data, updateData }: Props) => {
    const parsedData = React.useMemo(() => parse(data), [data]);

    const [time, setTime] = React.useState({ hour: parsedData.hour, minute: parsedData.minute });

    const displayHours = React.useMemo(() => timeIntoString(time.hour), [time.hour]);
    const displayMinutes = React.useMemo(() => timeIntoString(time.minute), [time.minute]);

    const updateTime = React.useCallback((newTime) => setTime(newTime), []);

    return (
        <div className="time-frame">
            <div className="time-frame__showtime">
                {displayHours}:{displayMinutes}
            </div>
            <div className="time-frame__control">
                <TimeList {...{ parsedData, time, updateData, updateTime }} type={TIME_TYPE.HOUR} />
                <TimeList {...{ parsedData, time, updateData, updateTime }} type={TIME_TYPE.MINUTE} />
            </div>
        </div>
    );
};

export default Time;
