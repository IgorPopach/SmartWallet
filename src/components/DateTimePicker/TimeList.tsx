import React from 'react';
import { DateTime } from 'luxon';

import { timeIntoString } from './Time';
import { TIME_TYPE } from './Time';

interface Time {
    hour: number;
    minute: number;
}

interface Props {
    type: typeof TIME_TYPE.HOUR | typeof TIME_TYPE.MINUTE;
    parsedData: DateTime;
    time: Time;
    updateData: (value: number) => void;
    updateTime: (newTime: Time) => void;
}

const TimeList = ({ type, parsedData, time, updateData, updateTime }: Props) => {
    const timeArr = React.useMemo(() => {
        if (type === TIME_TYPE.HOUR) {
            return new Array(24).fill(null).map((val, i) => i);
        }
        return new Array(60).fill(null).map((val, i) => i);
    }, []);

    const emptyItems = React.useMemo(() => new Array(2).fill(<div className="select-aria__item" />), []);

    const refs = React.useRef([]);
    if (refs.current.length !== timeArr.length) {
        refs.current = new Array(timeArr.length).fill(null).map((_, i) => refs.current[i] || React.createRef());
    }

    const list = React.useMemo(
        () =>
            timeArr.map((e, index) => {
                const handleClick = () => {
                    updateData(parsedData.set({ [type]: e }).toMillis());
                    updateTime({ ...time, [type]: e });
                    refs.current[e].current.scrollIntoView({
                        behavior: 'smooth',
                        block: 'center',
                    });
                };
                return (
                    <div
                        className="select-aria__item"
                        ref={refs.current[e]}
                        key={index}
                        role="option"
                        aria-selected={e === time[type as keyof Time]}
                        onClick={handleClick}
                    >
                        {timeIntoString(e)}
                    </div>
                );
            }),
        [time],
    );

    React.useEffect(() => {
        refs.current[time[type as keyof Time]].current.scrollIntoView({
            block: 'center',
        });
    }, []);

    return (
        <div role="listbox" className="time-frame__listbox">
            <div className="select-aria">
                {emptyItems}
                {list}
                {emptyItems}
            </div>
        </div>
    );
};

export default TimeList;
