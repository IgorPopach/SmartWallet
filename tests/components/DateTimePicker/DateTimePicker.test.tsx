import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import DateTimePicker from './../../../src/components/DateTimePicker/DateTimePicker';

type Props = React.ComponentProps<typeof DateTimePicker>;

describe('<DateTimePicker />', () => {
    const props: Props = {
        date: 123,
        handleClose: jest.fn(),
        onChange: jest.fn(),
        className: 'style',
    };

    let wrapper: ShallowWrapper<Props>;
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    // tslint:disable-next-line:no-any
    useStateSpy.mockImplementation((init: any) => [init, setState]);

    beforeAll(() => {
        wrapper = shallow(<DateTimePicker {...props} />);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('Should have correct layout', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Render initial component', () => {
        expect(wrapper.find('Calendar')).toHaveLength(1);
        expect(wrapper.find('Time')).toHaveLength(0);
    });

    it('Should call setTab with 1', () => {
        wrapper
            .find('div.datetime_btn')
            .at(1)
            .simulate('click');
        expect(setState).toHaveBeenCalledWith(1);
    });

    it('Should call setTab with 0', () => {
        wrapper
            .find('div.datetime_btn')
            .first()
            .simulate('click');
        expect(setState).toHaveBeenCalledWith(0);
    });
});
