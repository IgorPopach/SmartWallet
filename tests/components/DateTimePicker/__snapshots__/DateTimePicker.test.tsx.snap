// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateTimePicker /> Should have correct layout 1`] = `
<div
  className="datetime__frame style"
>
  <div
    className="datetime__control"
  >
    <div
      className="datetime_btn is-active"
      onClick={[Function]}
      updateData={[Function]}
    >
      Date
    </div>
    <div
      className="datetime_btn"
      onClick={[Function]}
    >
      Time
    </div>
  </div>
  <div
    className="datetime__body"
  >
    <Calendar
      data={123}
      updateData={[Function]}
    />
  </div>
  <div
    className="datetime_submit_btn"
    onClick={[MockFunction]}
  >
    Set
  </div>
</div>
`;
